zenoh-ext/src/net/querying_subscriber.rs

```diff
@@ -0,0 +1,483 @@
+            query_target: QueryTarget::default(),
+            query_consolidation: QueryConsolidation::default(),
+        }
+    }
+
+    /// Change the subscription reliability to Reliable.
+    pub fn reliable(mut self) -> Self {
+        self.info.reliability = Reliability::Reliable;
+        self
+    }
+
+    /// Change the subscription reliability to BestEffort.
+    pub fn best_effort(mut self) -> Self {
+        self.info.reliability = Reliability::BestEffort;
+        self
+    }
+
+    /// Change the subscription mode to Push.
+    pub fn push_mode(mut self) -> Self {
+        self.info.mode = SubMode::Push;
+        self.info.period = None;
+        self
+    }
+
+    /// Change the subscription mode to Pull with an optional Period.
+    pub fn pull_mode(mut self, period: Option<Period>) -> Self {
+        self.info.mode = SubMode::Pull;
+        self.info.period = period;
+        self
+    }
+    /// Change the resource key to be used for queries.
+    pub fn query_reskey(mut self, query_reskey: ResKey) -> Self {
+        self.query_reskey = query_reskey;
+        self
+    }
+
+    /// Change the predicate to be used for queries.
+    pub fn query_predicate(mut self, query_predicate: String) -> Self {
+        self.query_predicate = query_predicate;
+        self
+    }
+
+    /// Change the target to be used for queries.
+    pub fn query_target(mut self, query_target: QueryTarget) -> Self {
+        self.query_target = query_target;
+        self
+    }
+
+    /// Change the consolidation mode to be used for queries.
+    pub fn query_consolidation(mut self, query_consolidation: QueryConsolidation) -> Self {
+        self.query_consolidation = query_consolidation;
+        self
+    }
+}
+
+impl<'a> Future for QueryingSubscriberBuilder<'a> {
+    type Output = ZResult<QueryingSubscriber<'a>>;
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        Poll::Ready(QueryingSubscriber::new(Pin::into_inner(self).clone()))
+    }
+}
+
+impl<'a> ZFuture<ZResult<QueryingSubscriber<'a>>> for QueryingSubscriberBuilder<'a> {
+    fn wait(self) -> ZResult<QueryingSubscriber<'a>> {
+        QueryingSubscriber::new(self)
+    }
+}
+
+pub struct QueryingSubscriber<'a> {
+    conf: QueryingSubscriberBuilder<'a>,
+    subscriber: Subscriber<'a>,
+    receiver: QueryingSubscriberReceiver,
+}
+
+impl QueryingSubscriber<'_> {
+    fn new(conf: QueryingSubscriberBuilder<'_>) -> ZResult<QueryingSubscriber<'_>> {
+        // declare subscriber at first
+        let mut subscriber = conf
+            .session
+            .declare_subscriber(&conf.sub_reskey, &conf.info)
+            .wait()?;
+
+        let receiver = QueryingSubscriberReceiver::new(subscriber.receiver().clone());
+
+        let mut query_subscriber = QueryingSubscriber {
+            conf,
+            subscriber,
+            receiver,
+        };
+
+        // start query
+        query_subscriber.query()?;
```

Comment:
Maybe we need to add a timeout for the query. Otherwise the querying subscriber will block if one of the query target is under deadlock or live-lock. This will protect us until we don't allow to support manually asserted liveliness for queryables.

Reply:
Sure. But it would be simpler to rely on timeout implementation in the `zenoh::net::Session::query()`.
#106 created for this.